### ISSUE TITLE
StackLayout children rearrange themselves unexpectedly when their parent's size changes 

### ISSUE THREAD

**Issue by JackAnderson5** [kivy/kivy#4504]
Here is [the code ](https://gist.github.com/JackAnderson5/2c24c2bcf6da4bcee6219f1fabb16014).

I'm trying to populate a StackLayout with 5 buttons, so i set their `size_hint_x` to "one fifth" of their parent, or `1/5.`. This is equal to `0.2`. 

So, when parent width is `800` everything is fine; 5 buttons per row.

When the parent width is `801` each button has width `801*0.2`, that is `160.20000000000002` (floating point precision issues... the actual value should be `160.2`). The problem is that if we add 5 of those we get `160.20000000000002*5` which is `801.0000000000001`, which is greater than `801`. Therefor **i assume** StackLayout says "that exceeds my dimensions! Send it to the next raw."; results in 4 buttons per row.

In other words, floating point precision increases `width` or `height` by a tiny bit making children not fit sometimes. 

<i>Note that `801*(1/5.)` **!=** `801*1/5`, kivy gets the first one.</i>

**Comment by matham** [comment]
That seems like a problem that would be hard to solve generally.

**Comment by JackAnderson5** [comment]
Wouldn't simply rounding down the size of children solve it? 

e.g. `160.20000000000002` -> `160.2000000000000`

**Comment by matham** [comment]
Yeah, but is a significant amount? What works for your case will not work when the number of children is a lot higher. And rounding down arbitrarily the size of children by one will display incorrectly in other situations. 

I'm not sure if rounding will not effect how the user sees graphics when we display anything that is smaller than one pixel on the same line. Because sometimes it really does need to be displayed on the next row. Although I'm not sure how sensitive users are to these sub-pixel issues.

**Comment by JackAnderson5** [comment]
I'm not very familiar with kivy but.. 

> [...]will not work when the number of children is a lot higher

.. in order to be off by a single pixel in a whole StackLayout there would have to be trillions* of widgets on the same row of that StackLayout 

*(assuming we use something like `round(x, 12)` instead of `round(x, 3)`)

**Comment by matham** [comment]
Yeah, I guess you are right. Something like when `sum(children) - available_width < 1e-9` they are considered to be on the same line. A PR is welcome.

**Comment by JackAnderson5** [comment]
PR nearly ready, will do some benchmarking on performance of the 2 about solutions (rounding vs subtracting), and will PR.

**Comment by JackAnderson5** [comment]
[Benchmarking](https://gist.github.com/JackAnderson5/ae484a8f49cdda9aa9ecd9627a5e197d) indicates `round` is 15 times slower, so i used the subtraction.

WARNING: I didn't understand at all what StackLayout code does. It's `do_layout` is a 300SLOC method :P Instead i simply ran the debugger, replicated the resizing and searched where `160.20000000000002` occurred. I founded it, changed it, and tested it. Seems to be working fine.

**Comment by user5061** [comment]
Thinking about the subtraction (or rounding) solution.. this will not completely eliminate reordering-resizing in some specific conditions, but it will remove* it from obvious cases (e.g. expecting `size_hint_x=.2` on each widget in order to get 5 widgets on the same line of a StackLayout). 

I think you are right in saying this has no easy solution. I'm not confident with either of round or subtraction, but it's your call.

<sub>*Or... at least thats what i hope based on the fact i can't come up with an example that proves otherwise.</sub>
